streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```
@@ -265,7 +265,7 @@ private void verifyAllKVKeys(final StreamRunnable[] streamRunnables, final Kafka
                 public boolean conditionMet() {
                     try {
                         final StreamsMetadata metadata = streams.metadataForKey(storeName, key, new StringSerializer());
-                        if (metadata == null) {
+                        if (metadata == null || metadata.equals(StreamsMetadata.NOT_AVAILABLE)) {
```

Are users expected to check both conditions every time? Isn't this a bit error-prone?

---

They have a different meaning. `null` means that the `storeName` doesn't exist - perhaps it should throw an exception in this case. 
`StreamsMetadata.NOT_AVAILABLE` means the streams haven't (re-)initialized

---

Yeah, I figured they had different meanings, but seems a bit magical. :)

I would agree that the case where the store name doesn't exist should throw an exception if that is considered a user error (is there a good reason to call this method when the `storeName` doesn't exist?).

Aside from that, I'm not a fan of sentinel values because you have to know about them to use them properly. If you don't, you'll happily pass around the instance until some code tries to use something inside the sentinel that finally causes an error (probably the `HostInfo` in this case). Was this option chosen because the stream re-initialising is not considered an exceptional case?

---

People might call it with an invalid `storeName`, so it will never exist. I think it should throw in this case. @guozhangwang - thoughts?

Yep, we just had a long conversation about this on another PR.  Stream re-init is not really exceptional and people should back-off and re-try in these cases. Though, i could argue for throwing an exception here, too :-)

---

`Optional<StreamsMetadata>` would be a reasonable solution for the reinitialisation case, but we can't use it sadly.

---

Soon @ijuma, soon ;-)

---

@enothereska @ijuma @dguy OMG I love public API discussions :P

For `store name not exist`: this is not transient since the store names are determined by the time the topology is completely defined; i.e. if a store name does not exist now it will never exist in the future, in this case both throwing an exception (to be more restrictive) or returning null (like in a hash table) looks good to me.

For `metadata not available yet`: this is a transient state, and users should in this case just back-off and retry. If we decided to throw exception in the first case, then perhaps we can just return `null` in this case which do not users to learn about the sentinel value; if we decide to return `null` in the first case then in order to differentiate I'd suggest using a sentinel value.

In general I'd prefer to now throw exceptions for "retry scenarios" unless there is no other choices. Since the `try-catch` pattern looks worse than `if (null)` condition. But again this is purely subjective.